### PR TITLE
Add Plugin System to JDBI

### PR DESCRIPTION
**Pull request by stevenschlansker**
hi @qualidafial @brianm @arteam I am thinking about how we might make it easy to extend jdbi with a plugin like system.  Already, we have some obvious use cases -- Guava support without linking directly against Guava, SQLObject hooking into core, etc.

I have made a test branch here which adds support for automatic discovery of `JdbiPlugin`s via the classpath by a `ServiceLoader`.  This is little more than an idea at this point, but there are two in-tree use cases that are already interesting --
* Guava plugin can register itself with all DBI instances just by being present on classpath
* `HandleTrackerDBI` used in a test case can be a composable plugin, rather than needing to subclass `DBI`

What do you guys think of going down this direction?  I think it would be entirely optional, and the automatic registration should be used only for plugins which are transparent if not used.  For example adding support for custom types (Guava, joda-time) is OK as an automatic plugin, but replacing default transaction semantics would not be.

The diff is cluttered a bit because I converted DBI constructors to factory methods.  I'll try to annotate the interesting bits.

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/DBI.java [hunk 1]
```diff
@@ -62,28 +65,38 @@
...
-     * @param connectionFactory PrvidesJDBC connections to Handle instances
+     * @param connectionFactory Prvides JDBC connections to Handle instances
      */
-    public DBI(ConnectionFactory connectionFactory)
-    {
-        assert connectionFactory != null;
-        this.connectionFactory = connectionFactory;
+    public static DBI create(ConnectionFactory connectionFactory) {
+        DBI dbi = new DBI(connectionFactory);
+        dbi.installPlugins();
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/DBI.java]: By default, install all classpath plugins.

File: core/src/main/java/org/jdbi/v3/DBI.java [hunk 2]
```diff
@@ -114,9 +136,37 @@ public DBI(final String url, final Properties props)
...
+    public DBI installPlugin(JdbiPlugin plugin)
+    {
+        plugins.add(plugin);
+        return this;
+    }
+
+    public DBI resetPlugins()
     {
-        this(() -> DriverManager.getConnection(url, username, password));
+        plugins.clear();
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/DBI.java]: but you can get rid of them if you want total control

File: core/src/main/java/org/jdbi/v3/DBI.java [hunk 3]
```diff
@@ -198,6 +248,7 @@ public Handle open()
                                        MappingRegistry.copyOf(mappingRegistry),
                                        foreman.createChild(),
                                        collectorFactoryRegistry.createChild());
+            plugins.forEach(p -> p.customizeHandle(h));
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/DBI.java]: handle customization

File: core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.spi;
+
+import org.jdbi.v3.Handle;
+
+public interface JdbiPlugin {
+    default void customizeHandle(Handle handle) { }
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java]: everything is a default method, so we can add functionality later without breaking

File: guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ */
+package org.jdbi.v3.guava;
+
+import org.jdbi.v3.Handle;
+import org.jdbi.v3.spi.JdbiPlugin;
+
+public class GuavaPlugin implements JdbiPlugin {
+    @Override
+    public void customizeHandle(Handle handle) {
+        handle.registerCollectorFactory(GuavaCollectors.factory());
```
Comment (stevenschlansker) [inline guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java]: simplest useful plugin

File: guava/src/main/resources/META-INF/services/org.jdbi.v3.spi.JdbiPlugin [hunk 6]
```diff
@@ -0,0 +1 @@
+org.jdbi.v3.guava.GuavaPlugin
```
Comment (stevenschlansker) [inline guava/src/main/resources/META-INF/services/org.jdbi.v3.spi.JdbiPlugin]: easy to register with the `ServiceLoader`

File: guava/src/test/java/org/jdbi/v3/guava/TestGuavaCollectors.java [hunk 7]
```diff
@@ -70,8 +70,6 @@ public void addData() {
             db.getSharedHandle().execute("insert into something(name, intValue) values (?, ?)", Integer.toString(i), i);
             expected.add(i);
         }
-
-        db.getSharedHandle().registerCollectorFactory(GuavaCollectors.factory());
```
Comment (stevenschlansker) [inline guava/src/test/java/org/jdbi/v3/guava/TestGuavaCollectors.java]: no need to register by hand now

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java [hunk 8]
```diff
@@ -268,19 +256,13 @@ public Something map(int index, ResultSet r, StatementContext ctx) throws SQLExc
         }
     }
 
-    public static class HandleTrackerDBI extends DBI
+    static class HandleTracker implements JdbiPlugin
```
Comment (stevenschlansker) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java]: now this can be composed in, rather than relying on subclassing

File: core/src/main/java/org/jdbi/v3/DBI.java (hunk 3 above)
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/DBI.java]: Minor nit: could be `plugins.forEach(JdbiPlugin::customizeHandle)`

Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/DBI.java]: If only... You have to bind 'h' :)

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBindAutomaticNames.java [hunk 9]
```diff
@@ -13,52 +13,41 @@
...
-
-    @Override
-    public void tearDown() throws Exception
-    {
-        handle.execute("drop table something");
-        handle.close();
+        handle = db.getSharedHandle();
     }
 
+    @Test
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBindAutomaticNames.java]: wut

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestMapBinder.java [hunk 10]
```diff
@@ -24,33 +26,33 @@
...
     private Handle handle;
 
-    @Override
+    @Before
     public void setUp() throws Exception
     {
         JdbcDataSource ds = new JdbcDataSource();
         ds.setURL("jdbc:h2:mem:test");
-        dbi = new DBI(ds);
+        dbi = DBI.create(ds);
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestMapBinder.java]: Should use `H2DatabaseRule` here too.

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestMixinInterfaces.java [hunk 11]
```diff
@@ -42,7 +42,7 @@ public void setUp() throws Exception
     {
         JdbcDataSource ds = new JdbcDataSource();
         ds.setURL(String.format("jdbc:h2:mem:%s;MVCC=TRUE", UUID.randomUUID()));
-        dbi = new DBI(ds);
+        dbi = DBI.create(ds);
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestMixinInterfaces.java]: Should use H2DatabaseRule

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestNewApiOnDbiAndHandle.java [hunk 12]
```diff
@@ -40,7 +40,7 @@ public void setUp() throws Exception
     {
         JdbcDataSource ds = new JdbcDataSource();
         ds.setURL("jdbc:h2:mem:" + UUID.randomUUID());
-        dbi = new DBI(ds);
+        dbi = DBI.create(ds);
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestNewApiOnDbiAndHandle.java]: Should use H2DatabaseRule

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestPositionalBinder.java [hunk 13]
```diff
@@ -17,39 +17,33 @@
...
     @Before
     public void setUp() throws Exception {
-        JdbcDataSource ds = new JdbcDataSource();
-        ds.setURL("jdbc:h2:mem:" + UUID.randomUUID());
-        DBI dbi = new DBI(ds);
-        handle = dbi.open();
+        handle = db.getSharedHandle();
         somethingDao = SqlObjectBuilder.attach(handle, SomethingDao.class);
 
+        handle.execute("drop table something");
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestPositionalBinder.java]: If this test is isolated, why should there be any tables to drop?

File: core/src/main/java/org/jdbi/v3/DBI.java (hunk 1 above)
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/DBI.java]: This is the only thing I object to. I'm not a fan of Spring-style magic--by which I mean stuff just happens because a JAR is on the classpath (or sometime next year, in the Java 9 module graph).

JDBI has thus far done a great job of keeping the magic "visible." I would rather force users to make a single call to `DBI.installPlugins()` so that the ServiceLoader stuff is truly opt-in.

**Comment by qualidafial** [comment]
This is a fantastic addition! :+1:

**Comment by stevenschlansker** [comment]
Fixed up most of the comments, everything else can be fixed in post ;)

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestPositionalBinder.java (hunk 13 above)
Comment (stevenschlansker) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestPositionalBinder.java]: the rule creates it automagically, and this test uses a conflicting table definition, so we drop and recreate it.  not the best, but should be fixed as a separate PR

**Comment by arteam** [comment]
This looks like a great framework for handle customizing. It could useful for some specific database features or Java containers.

**Comment by stevenschlansker** [comment]
Yeah!  My thoughts exactly.

**Comment by brianm** [comment]
I realize I am late to the party, but I much prefer explicit registration rather than "behavior magically changes because something was added to the classpath."

File: core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.spi;
+
+import org.jdbi.v3.Handle;
+
+public interface JdbiPlugin {
+    default Handle customizeHandle(Handle handle) { return handle; }
```
Comment (brianm) [inline core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java]: Customize statement?

Comment (brianm) [inline core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java]: Customize DBI?

Comment (brianm) [inline core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java]: I realize customizing a statement can be done by providing a wrapper handle, but that is kind of inconvenient compared to making it first class on the plugin. It also implies separate methods for customizing queries, statements, batches, etc, so as to handle all the downcast cases from statement correctly.

File: guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package org.jdbi.v3.guava;
+
+import org.jdbi.v3.Handle;
+import org.jdbi.v3.spi.JdbiPlugin;
+
+public class GuavaPlugin implements JdbiPlugin {
+    @Override
+    public Handle customizeHandle(Handle handle) {
+        handle.registerCollectorFactory(GuavaCollectors.factory());
```
Comment (brianm) [inline guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java]: I would expect this to also register handlers for passing in Guava's Optional<?> types for binding (ie, if none then bind null, if some then bind value).

File: core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java (hunk 14 above)
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/spi/JdbiPlugin.java]: Yes, both of these things are reasonable to add.

**Comment by stevenschlansker** [comment]
Yeah, the merged version removes the magic auto-registration unless explicitly asked.

File: guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java (hunk 15 above)
Comment (stevenschlansker) [inline guava/src/main/java/org/jdbi/v3/guava/GuavaPlugin.java]: Hm, could add that, although at this point Guava's optional type should be discouraged for the Java 8 version :)
